examples/pytorch/adaptive_sampling/adaptive_sampling.py

```diff
@@ -0,0 +1,475 @@
+test_h = torch.tensor(data.features[test_nodes], dtype=torch.float64)
+all_h = torch.tensor(data.features, dtype=torch.float64)
+train_nodes = torch.tensor(train_nodes)
+test_nodes = torch.tensor(test_nodes)
+y_train = torch.tensor(data.labels[train_nodes])
+y_test = torch.tensor(data.labels[test_nodes])
+##configuration
+n_epoch=300
+lamb = 0.5
+lr = 1e-3
+weight_decay = 5e-4
+input_size = h.shape[1]
+hidden_size = 16
+output_size = data.num_labels
+##sample size for each layer during training
+layer_sizes = [256, 256]
+batch_size = 256
+##sample size for each layer during test
+test_batch_size = 64
+avg_degree = 8
+test_layer_sizes = [test_batch_size * avg_degree for _ in range(2)]
+
+
+class Sampler(object):
+    def __init__(self, graph):
+        self.graph = graph
+
+    def __len__(self):
+        raise NotImplementedError
+
+    def __iter__(self):
+        """Iterator
+
+        The iterator must return a tuple at a time.  The first element must be an array of node IDs from which
+        NodeFlow is grown.  Other elements can be arbitrary auxiliary data correspond to the current minibatch.
+        """
+        raise NotImplementedError
+
+
+class NodeSampler(Sampler):
+    """Minibatch sampler that samples batches of nodes uniformly from the given graph and list of seeds.
+    """
+
+    def __init__(self, graph, seeds, batch_size):
+        super().__init__(graph)
+        self.seeds = seeds
+        self.batch_size = batch_size
+
+    def __len__(self):
+        return len(self.seeds) // self.batch_size
+
+    def __iter__(self):
+        """Returns
+        (1) The seed node IDs, for NodeFlow generation,
+        (2) Indices of the seeds in the original seed array, as auxiliary data.
+        """
+        batches = torch.randperm(len(self.seeds)).split(self.batch_size)
+        for i in range(len(self)):
+            if len(batches[i]) < self.batch_size:
+                break
+            else:
+                yield self.seeds[batches[i]], batches[i]
+
+
+def create_nodeflow(layer_mappings, block_mappings, block_aux_data, rel_graphs, seed_map):
+    hg = dgl.hetero_from_relations(rel_graphs)
+    hg.layer_mappings = layer_mappings
+    hg.block_mappings = block_mappings
+    hg.block_aux_data = block_aux_data
+    hg.seed_map = seed_map
+    return hg
+
+
+def normalize_adj(adj):
+    """Symmetrically normalize adjacency matrix."""
+    adj = ssp.eye(adj.shape[0]) + adj
+    adj = ssp.coo_matrix(adj)
+    rowsum = np.array(adj.sum(1))
+    d_inv_sqrt = np.power(rowsum, -0.5).flatten()
+    d_inv_sqrt[np.isinf(d_inv_sqrt)] = 0.
+    d_mat_inv_sqrt = ssp.diags(d_inv_sqrt)
+    return adj.dot(d_mat_inv_sqrt).transpose().dot(d_mat_inv_sqrt)
+
+
+class Generator(object):
+    def __init__(self, sampler=None, num_workers=0):
+        self.sampler = sampler
+        self.num_workers = num_workers
+
+        if num_workers > 0:
+            raise NotImplementedError('multiprocessing')
```

Comment:
Based on the same rationale, you could

(1) combine `Generator`, `IterativeGenerator` and `AdaptGenerator` into one class, and
(2) remove the `num_workers` argument